### ISSUE TITLE
question: Accept Basic auth if found in request headers

### ISSUE THREAD

**Issue by connordeckers** [greenpau/caddy-authorize#90]
> A clear and concise description of what you want to accomplish.

When using this for authorizing a `reverse_proxy`, some services are built to work with `Basic` auth (inc. additional services that integrate with said service).

For example, Transmission only supports Basic auth as a login system (or none at all). As this is the expectation, any services that interact with Transmission will expect that, if they authenticate, it's via `Basic` auth. 

As I can't have say "sign in to this URL, then go over to that URL" to these various integrations, using `<<auth-endpoint>>/login/basic/<<realm>>` isn't a suitable solution, nor is it suitable to remove the authentication altogether.

For context, I'm using the `local` implementation, with a working user available in the `.json` file already.

Is there any way that I can accept Basic auth if found in the Headers?

Cheers!

**Comment by greenpau** [comment]
> As I can't have say "sign in to this URL, then go over to that URL" to these various integrations, using <<auth-endpoint>>/login/basic/<<realm>> isn't a suitable solution, nor is it suitable to remove the authentication altogether.

@connordeckers, kudoz for reading and educating yourself on the subject 👍

> For example, Transmission only supports Basic auth as a login system (or none at all). As this is the expectation, any services that interact with Transmission will expect that, if they authenticate, it's via Basic auth.

I could add basic authentication, but it would not send [`WWW-Authenticate: Basic`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate) to the requester.

Would work for your use case?

Alternatively, we could parse User-Agent and determine the clients that do not support redirects.

**Comment by connordeckers** [comment]
Apologies for the massive delay! I ended up making a bit of a hack, using the following block:
```caddyfile
handle {
    authorize {
        bypass uri prefix /transmission/rpc
    }

    reverse_proxy http://transmission:9091
}
```

It's not entirely ideal, but it works for the moment. If there was a way to use authentication (so that it's not set up so that anyone who knows it's there can interact with that URL), that would be awesome!

The way I'm envisaging it is:

- User navigates to URL in browser
- Caddy notes that no authentication is defined, goes through regular HTTP/HTML auth process
- User continues on their merry way
---
- User sends HTTP request with `Auth: Basic <<base64 hash of username and password>>`
- Caddy-Auth can test that, go "yep, you're who we think you are, carry on", allows request through
- User has "transparently" authenticated via HTTP header, without being redirected to HTML login page


Does that seem feasible? I've not had a dig into the code (quite green when it comes to Golang), but I can imagine that being able to login via Basic auth would be useful for API services being proxied behind Caddy.

Apologies for telling you how to work on your own project; I love the work you've done!! :)

**Comment by greenpau** [comment]
> Does that seem feasible? I've not had a dig into the code (quite green when it comes to Golang), but I can imagine that being able to login via Basic auth would be useful for API services being proxied behind Caddy.

@connordeckers , please paste your config here (or create a gist with the config) and I will show you how "wrangle" it!

**Comment by greenpau** [comment]
@connordeckers , also, add [`caddy-trace`](https://github.com/greenpau/caddy-trace) plugin and provide an output for the Transmission client connection. You will see the User-Agent in there.

**Comment by greenpau** [comment]
@connordeckers , I am working on basic auth and api keys integration with `authorize` plugin. Will keep you posted here.

**Comment by greenpau** [comment]
@connordeckers , this is now available with [caddy-auth-portal v1.4.37](https://github.com/greenpau/caddy-auth-portal/releases/tag/v1.4.37).

Please see https://gist.github.com/greenpau/b7fbdc3641a8a51a5d6f47e1c7af641b

Note the `with basic auth realm local` directive.


Once configured, you should be able to access endpoints with Basic Authentication, e.g.:

```bash
AUTH_HEADER=$(echo -ne 'webadmin:asd3kldkjld23HJh33jDnx!@1' | base64 --wrap 0)
curl -v -H "Authorization: Basic ${AUTH_HEADER}" https://assetq.myfiosgateway.com:8443/
```

Please test 😃

**Comment by connordeckers** [comment]
@greenpau Thank you so much! Sorry for the significant delay in replying; my server crashed, and then christmas happened.
I will most certainly test this out, and let you know how it all goes!

Thank you again for being so incredibly active on this! :)
